add WithHeader and convert NewClient to use options

jabielecki opened the pull request:
### Primary Motivation

Described in internal ticket ENG-20050. This change is related to one of EM projects.

### Description

~Add yet another constructor in order to provide means for callers to insert custom headers into every request. For example, if they need a third header in addition to the typical X-Auth-Key and X-Consumer-Token, they can use `NewClientWithHeader()` to obtain that.~

Add the ClientOpt, just classic functional options:
https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis

The immediate need is to provide means for callers to insert custom headers into every request. For example, if callers need a third header in addition to the typical X-Auth-Key and X-Consumer-Token pair, they can now use WithHeader.

The existing possibility of customizing headers per every request with DoRequestWithHeaders has the problem that for an existing code base using packngo.DoRequest heavily, it becomes hard to add a header to every request without a massive code modification.

### Compatibility

I modified the constructor NewClient in a **backward-compatible way** to use the functional options. The small difference in logic is that if the builtin default base API URL turns out to be non-parseable:

  - now NewClient will return: nil Client, appropriate error
  - previously NewClient returned: nil Client, nil error

Also the order of reading env variables changed, which means that from now on majority of the constructor code can benefit from
Client.debug (`$PACKNGO_DEBUG`). Previously it was known quite late into the construction.

### Unit Test Simplification

Remove duplicated code: the function claiming to be a "mockX" is in fact "exactlyDuplicatedX". Do not duplicate code from packngo.go with the sole purpose of passing a test through it. The test's goal is to warn about bad changes to the _real_ code - focus the test on that goal.

displague commented on packngo.go:
```diff
@@ -355,6 +355,13 @@ func NewClientWithAuth(consumerToken string, apiKey string, httpClient *http.Cli
+// NewClientWithHeader returns a Client which for every new request adds given HTTP header set. Otherwise it behaves
+// like NewClientWithBaseURL. The headers X-Auth-Token, X-Consumer-Token, User-Agent will be ignored if provided
+// in the set.
+func NewClientWithHeader(consumerToken string, apiKey string, httpClient *http.Client, apiBaseURL string, header http.Header) (*Client, error) {
```
I've considered adding other constructors - adding a new one for each new field we add later may be messy.

What are your thoughts on changing `NewClient()` to `NewClient(...ClientOpt)` where `ClientOpt` is `func (*Client) error`.  We can then add `ClientOpt` helpers for headers, base URL, Auth, UserAgent, etc.

jabielecki replied:
Hi @displague , do you mean - as part of this PR?
I'm familiar with the convention and I can add it, my only concern is that's not super-discoverable for users. (A comment would fix that.)

The other fancy option are chained builders:

```
c := NewClient()
```

```
c := NewClient().WithHeaders(h)
```

```
c := NewClient().
        WithHeaders().
        WithSomethingElse().
        WithYetAnotherThingie()
```

What do you think?

displague replied:
It's good that you're considering discoverability.

This particular pattern won't against NewClient which returns `(*Client,error)` today.

jabielecki replied:
Added :smile:

jabielecki commented:
Hi @displague , in PR top document I've rewritten much of the "Description" section and added a new "Compatibility" section. Also, rewritten most of my code. Could you review, please?

jabielecki commented:
Added private bool field `Client.apiKeySet`.

While it adds a bit of complexity, it's a price to keep backward compatibility for users who emitted an empty X-Auth-Token
(verbatim `X-Auth-Token: ` and end of line). Could happen in tests and whatnot.

ocobleseqx reviewed (approved):
good job!
